narwhal/primary/src/core.rs

```diff
@@ -180,6 +179,51 @@ impl Core {
 
         self.highest_received_round = last_round_number;
         self.highest_processed_round = last_round_number;
+
+        // Get latest header from all peers and process it.
+        let peers: Vec<NetworkPublicKey> = self
+            .committee
+            .others_primaries(&self.name)
+            .into_iter()
+            .map(|(_, _, network_key)| network_key)
+            .collect();
+
+        let network = P2pNetwork::new(self.network.network());
+        let mut header_futures = FuturesUnordered::new();
+        let request = LatestHeaderRequest {};
+        for peer in peers.iter() {
+            let network = network.network();
+            let request = request.clone();
+
+            header_futures.push(async move {
+                let _ = &network;
+                network.get_latest_header(peer, request).await
+            });
+        }
+        while let Some(res) = header_futures.next().await {
+            match res {
+                Ok(response) => {
+                    if let Some(header) = response.header {
+                        let result = self.process_header(&header).await;
+                        if result.is_err() {
+                            error!(
+                                "error on recovery path processing latest header: {:?}",
+                                result.err()
+                            );
+                        }
```

Comment:
nit: alternatively you could do:
```
if let Err(err) = self.process_header(&header).await {
     error!("error on recovery path processing latest header: {:?}", err);
}
```

Reply:
On another note I a bit skeptical on putting for immediate processing the header we just fetched. If we have nodes that are slow to respond this `while loop` here might take some time finish while the core's `run` loop has not started yet to process any fetched certificates etc.  Maybe I would look to finish with all the fetching first and then put those for processing. After all we can't continue in Core unless we have finished with fetching the headers.

Reply:
Even better than blocking on all fetches, which stops Core from processing anything until all RPCs have finished, maybe chain the `get_latest_header` call and the `process_header` calls together with `and_then` up at line 200?

Reply:
Though I'm realizing that might not work because Core probably doesn't support concurrent calls to process_header... in which case what Laura has already might be best? (At least the `FuturesUnordered` will yield results as fast as remote nodes can return them, so I'm not sure if blocking on all fetch RPCs first buys us anything unless we're worried that it might not even get to *start* all the RPCs before the first header is back and being processed?)

Reply:
I like this idea

Reply:
@aschran my main concern is that by calling `process_header` we know that it's possible to kick off downstream processes that end up feeding back to Core. Also the `recover` method will not return until we have finished fetching all the headers taking into account timeouts . So in a worst case scenario it can take up to 15 seconds (with current timeout setting) for the `recover` method to return and then bootstrap the core `run` loop. Now within those `15 seconds` other components will do work which will not be consumed by Core.